File: clients/java/src/Pipeline.java
package com.fastkv.client;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * A pipeline that buffers multiple FastKV commands and sends them in a single round-trip.
 * <p>
 * Usage pattern:
 * <pre>
 *   try (Pipeline p = client.pipeline()) {
 *       p.set("k1", "v1");
 *       p.set("k2", "v2");
 *       p.get("k1");
 *       p.get("k2");
 *       List&lt;Object&gt; results = p.sync();
 *   }
 * </pre>
 * <p>
 * The pipeline opens its own dedicated socket connection so that no other threads
 * interleave commands on the same connection. The connection is closed when the
 * pipeline is closed.
 * <p>
 * <b>Thread safety:</b> Pipeline is NOT thread-safe. It is intended to be used by
 * a single thread. Each {@link #sync()} call flushes all buffered commands and
 * reads all their responses.
 */
public class Pipeline implements AutoCloseable {

    private final String host;
    private final int port;
    private final int connectTimeout;
    private final int soTimeout;

    private Socket socket;
    private OutputStream out;
    private InputStream in;

    /** Buffer of encoded RESP command byte arrays. */
    private final List<byte[]> commandBuffer = new ArrayList<>();

    /** Track which commands expect a boolean response (SET NX/XX, EXPIRE, etc.). */
    private final List<ResponseConverter> converters = new ArrayList<>();

    Pipeline(String host, int port, int connectTimeout, int soTimeout) {
        this.host = host;
        this.port = port;
        this.connectTimeout = connectTimeout;
        this.soTimeout = soTimeout;
    }

    // ──────────────────────────────────────────────────────────────────────
    // Connection management
    // ──────────────────────────────────────────────────────────────────────

    private void ensureConnected() {
        try {
            if (socket == null || socket.isClosed()) {
                socket = new Socket();
                socket.setTcpNoDelay(true);
                socket.setSoTimeout(soTimeout);
                socket.connect(new java.net.InetSocketAddress(host, port), connectTimeout);
                out = new BufferedOutputStream(socket.getOutputStream(), 8192);
                in = new BufferedInputStream(socket.getInputStream(), 8192);
            }
        } catch (IOException e) {
            throw new FastKVConnectionException("Failed to connect to " + host + ":" + port, e);
        }
    }

    private void writeCommand(byte[] encoded) {
        ensureConnected();
        try {
            out.write(encoded);
        } catch (IOException e) {
            throw new FastKVConnectionException("Failed to write command to pipeline", e);
        }
    }

    // ──────────────────────────────────────────────────────────────────────
    // ResponseConverter – describes how to convert a raw RESP reply
    // ──────────────────────────────────────────────────────────────────────

    @FunctionalInterface
    private interface ResponseConverter {
        Object convert(Object raw);
    }

    private static final ResponseConverter PASSTHROUGH = raw -> raw;
    private static final ResponseConverter TO_BOOLEAN = raw -> toBoolean(raw);
    private static final ResponseConverter TO_LONG = raw -> toLong(raw);

    // ──────────────────────────────────────────────────────────────────────
    // Flushing & syncing
    // ──────────────────────────────────────────────────────────────────────

    /**
     * Flushes all buffered commands to the server and reads all responses.
     *
     * @return a list of decoded responses, one per buffered command (in order)
     */
    public List<Object> sync() {
        ensureConnected();
        try {
            out.flush();
        } catch (IOException e) {
            throw new FastKVConnectionException("Failed to flush pipeline commands", e);
        }

        int count = commandBuffer.size();
        List<Object> results = new ArrayList<>(count);
        for (int i = 0; i < count; i++) {
            Object raw = RespDecoder.decode(in);
            results.add(converters.get(i).convert(raw));
        }
        commandBuffer.clear();
        converters.clear();
        return results;
    }

    // ──────────────────────────────────────────────────────────────────────
    // Core commands
    // ──────────────────────────────────────────────────────────────────────

    public Pipeline ping() {
        return addCommand("PING");
    }

    public Pipeline echo(String message) {
        return addCommand("ECHO", message);
    }

    public Pipeline info() {
        return addCommand("INFO");
    }

    public Pipeline dbsize() {
        return addCommand("DBSIZE");
    }

    // ──────────────────────────────────────────────────────────────────────
    // String commands
    // ──────────────────────────────────────────────────────────────────────

    public Pipeline set(String key, String value) {
        return addCommand("SET", key, value);
    }

    public Pipeline setEx(String key, String value, long seconds) {
        return addCommand("SET", key, value, "EX", String.valueOf(seconds));
    }

    public Pipeline setPx(String key, String value, long millis) {
        return addCommand("SET", key, value, "PX", String.valueOf(millis));
    }

    public Pipeline setNx(String key, String value) {
        return addCommand("SET", key, value, "NX").withConverter(TO_BOOLEAN);
    }

    public Pipeline setXx(String key, String value) {
        return addCommand("SET", key, value, "XX").withConverter(TO_BOOLEAN);
    }

    public Pipeline get(String key) {
        return addCommand("GET", key);
    }

    public Pipeline del(String... keys) {
        String[] args = new String[keys.length + 1];
        args[0] = "DEL";
        System.arraycopy(keys, 0, args, 1, keys.length);
        return addCommand(args).withConverter(TO_LONG);
    }

    public Pipeline exists(String... keys) {
        String[] args = new String[keys.length + 1];
        args[0] = "EXISTS";
        System.arraycopy(keys, 0, args, 1, keys.length);
        return addCommand(args).withConverter(TO_LONG);
    }

    public Pipeline incr(String key) {
        return addCommand("INCR", key).withConverter(TO_LONG);
    }

    public Pipeline decr(String key) {
        return addCommand("DECR", key).withConverter(TO_LONG);
    }

    public Pipeline incrBy(String key, long delta) {
        return addCommand("INCRBY", key, String.valueOf(delta)).withConverter(TO_LONG);
    }

    public Pipeline decrBy(String key, long delta) {
        return addCommand("DECRBY", key, String.valueOf(delta)).withConverter(TO_LONG);
    }

    public Pipeline append(String key, String value) {
        return addCommand("APPEND", key, value).withConverter(TO_LONG);
    }

    public Pipeline strlen(String key) {
        return addCommand("STRLEN", key).withConverter(TO_LONG);
    }

    public Pipeline getRange(String key, int start, int end) {
        return addCommand("GETRANGE", key, String.valueOf(start), String.valueOf(end));
    }

    public Pipeline setRange(String key, int offset, String value) {
        return addCommand("SETRANGE", key, String.valueOf(offset), value).withConverter(TO_LONG);
    }

    public Pipeline mset(Map<String, String> pairs) {
        String[] args = new String[pairs.size() * 2 + 1];
        args[0] = "MSET";
        int i = 1;
        for (Map.Entry<String, String> e : pairs.entrySet()) {
            args[i++] = e.getKey();
            args[i++] = e.getValue();
        }
        return addCommand(args);
    }

    public Pipeline mget(String... keys) {
        String[] args = new String[keys.length + 1];
        args[0] = "MGET";
        System.arraycopy(keys, 0, args, 1, keys.length);
        return addCommand(args);
    }

    // ──────────────────────────────────────────────────────────────────────
    // TTL commands
    // ──────────────────────────────────────────────────────────────────────

    public Pipeline expire(String key, long seconds) {
        return addCommand("EXPIRE", key, String.valueOf(seconds)).withConverter(TO_BOOLEAN);
    }

    public Pipeline ttl(String key) {
        return addCommand("TTL", key).withConverter(TO_LONG);
    }

    public Pipeline pttl(String key) {
        return addCommand("PTTL", key).withConverter(TO_LONG);
    }

    public Pipeline persist(String key) {
        return addCommand("PERSIST", key).withConverter(TO_BOOLEAN);
    }

    // ──────────────────────────────────────────────────────────────────────
    // Hash commands
    // ──────────────────────────────────────────────────────────────────────

    public Pipeline hset(String key, String field, String value) {
        return addCommand("HSET", key, field, value).withConverter(TO_BOOLEAN);
    }

    public Pipeline hget(String key, String field) {
        return addCommand("HGET", key, field);
    }

    public Pipeline hdel(String key, String... fields) {
        String[] args = new String[fields.length + 2];
        args[0] = "HDEL";
        args[1] = key;
        System.arraycopy(fields, 0, args, 2, fields.length);
        return addCommand(args).withConverter(TO_LONG);
    }

    public Pipeline hgetAll(String key) {
        return addCommand("HGETALL", key);
    }

    public Pipeline hexists(String key, String field) {
        return addCommand("HEXISTS", key, field).withConverter(TO_BOOLEAN);
    }

    public Pipeline hlen(String key) {
        return addCommand("HLEN", key).withConverter(TO_LONG);
    }

    public Pipeline hkeys(String key) {
        return addCommand("HKEYS", key);
    }

    public Pipeline hvals(String key) {
        return addCommand("HVALS", key);
    }

    public Pipeline hmget(String key, String... fields) {
        String[] args = new String[fields.length + 2];
        args[0] = "HMGET";
        args[1] = key;
        System.arraycopy(fields, 0, args, 2, fields.length);
        return addCommand(args);
    }

    public Pipeline hmset(String key, Map<String, String> fields) {
        String[] args = new String[fields.size() * 2 + 2];
        args[0] = "HMSET";
        args[1] = key;
        int i = 2;
        for (Map.Entry<String, String> e : fields.entrySet()) {
            args[i++] = e.getKey();
            args[i++] = e.getValue();
        }
        return addCommand(args);
    }

    // ──────────────────────────────────────────────────────────────────────
    // List commands
    // ──────────────────────────────────────────────────────────────────────

    public Pipeline lpush(String key, String... elements) {
        String[] args = new String[elements.length + 2];
        args[0] = "LPUSH";
        args[1] = key;
        System.arraycopy(elements, 0, args, 2, elements.length);
        return addCommand(args).withConverter(TO_LONG);
    }

    public Pipeline rpush(String key, String... elements) {
        String[] args = new String[elements.length + 2];
        args[0] = "RPUSH";
        args[1] = key;
        System.arraycopy(elements, 0, args, 2, elements.length);
        return addCommand(args).withConverter(TO_LONG);
    }

    public Pipeline lpop(String key) {
        return addCommand("LPOP", key);
    }

    public Pipeline lpop(String key, int count) {
        return addCommand("LPOP", key, String.valueOf(count));
    }

    public Pipeline rpop(String key) {
        return addCommand("RPOP", key);
    }

    public Pipeline rpop(String key, int count) {
        return addCommand("RPOP", key, String.valueOf(count));
    }

    public Pipeline lrange(String key, int start, int stop) {
        return addCommand("LRANGE", key, String.valueOf(start), String.valueOf(stop));
    }

    public Pipeline llen(String key) {
        return addCommand("LLEN", key).withConverter(TO_LONG);
    }

    public Pipeline lindex(String key, int index) {
        return addCommand("LINDEX", key, String.valueOf(index));
    }

    public Pipeline lrem(String key, int count, String element) {
        return addCommand("LREM", key, String.valueOf(count), element).withConverter(TO_LONG);
    }

    public Pipeline ltrim(String key, int start, int stop) {
        return addCommand("LTRIM", key, String.valueOf(start), String.valueOf(stop));
    }

    public Pipeline lset(String key, int index, String element) {
        return addCommand("LSET", key, String.valueOf(index), element);
    }

    // ──────────────────────────────────────────────────────────────────────
    // Internal helpers
    // ──────────────────────────────────────────────────────────────────────

    private Pipeline addCommand(String... args) {
        byte[] encoded = RespEncoder.encode(args);
        writeCommand(encoded);
        commandBuffer.add(encoded);
        converters.add(PASSTHROUGH);
        return this;
    }

    private Pipeline withConverter(ResponseConverter converter) {
        // Replace the last inserted converter
        converters.set(converters.size() - 1, converter);
        return this;
    }

    private static Long toLong(Object raw) {
        if (raw == RespDecoder.NULL) return null;
        if (raw instanceof Long) return (Long) raw;
        if (raw instanceof String) {
            try {
                return Long.parseLong((String) raw);
            } catch (NumberFormatException e) {
                throw new FastKVResponseException("Expected integer response, got: " + raw);
            }
        }
        throw new FastKVResponseException("Expected integer response, got: " + raw);
    }

    private static Boolean toBoolean(Object raw) {
        if (raw == RespDecoder.NULL) return null;
        if (raw instanceof Long) return ((Long) raw) == 1L;
        if (raw instanceof String) return "OK".equalsIgnoreCase((String) raw);
        throw new FastKVResponseException("Expected boolean response, got: " + raw);
    }

    // ──────────────────────────────────────────────────────────────────────
    // Close
    // ──────────────────────────────────────────────────────────────────────

    /**
     * Closes the pipeline's dedicated connection. Any buffered but un-synced
     * commands are discarded.
     */
    @Override
    public void close() {
        commandBuffer.clear();
        converters.clear();
        if (socket != null) {
            try {
                socket.close();
            } catch (IOException ignored) {
            }
            socket = null;
            out = null;
            in = null;
        }
    }

}

File: fastkv/clients/rust/src/pipeline.rs
//! FastKV async client — pipeline support.
//!
//! Pipeline batches multiple commands into a single write, then reads
//! all replies sequentially.

use crate::resp::{encode_command, read_reply, Error, Reply};
use tokio::io::BufReader;
use tokio::net::TcpStream;

/// A pipeline that collects commands and sends them in a single batch.
///
/// # Example
/// ```no_run
/// use fastkv_client::Client;
///
/// #[tokio::main]
/// async fn main() -> Result<(), Box<dyn std::error::Error>> {
///     let mut c = Client::connect("127.0.0.1", 8379).await?;
///     let mut pipe = c.pipeline();
///     pipe.set("k1", "v1");
///     pipe.set("k2", "v2");
///     pipe.get("k1");
///     let res = pipe.execute().await?;
///     assert_eq!(res.string(2)?, Some("v1".into()));
///     # Ok(())
/// }
/// ```
pub struct Pipeline<'a> {
    stream: &'a mut TcpStream,
    bufs: Vec<Vec<u8>>,
}

impl<'a> Pipeline<'a> {
    pub(crate) fn new(stream: &'a mut TcpStream) -> Self {
        Self {
            stream,
            bufs: Vec::new(),
        }
    }

    // ── String commands ──────────────────────────────────────────────────

    /// SET key value
    pub fn set(&mut self, key: &str, value: &str) {
        self.command(&[b"SET", key.as_bytes(), value.as_bytes()]);
    }

    /// GET key
    pub fn get(&mut self, key: &str) {
        self.command(&[b"GET", key.as_bytes()]);
    }

    /// DEL key [key ...]
    pub fn del(&mut self, keys: &[&str]) {
        let mut args: Vec<&[u8]> = vec![b"DEL"];
        for k in keys {
            args.push(k.as_bytes());
        }
        self.command(&args);
    }

    /// INCR key
    pub fn incr(&mut self, key: &str) {
        self.command(&[b"INCR", key.as_bytes()]);
    }

    /// INCRBY key n
    pub fn incr_by(&mut self, key: &str, n: i64) {
        self.command(&[b"INCRBY", key.as_bytes(), n.to_string().as_bytes()]);
    }

    /// DECR key
    pub fn decr(&mut self, key: &str) {
        self.command(&[b"DECR", key.as_bytes()]);
    }

    /// APPEND key value
    pub fn append(&mut self, key: &str, value: &str) {
        self.command(&[b"APPEND", key.as_bytes(), value.as_bytes()]);
    }

    /// MGET key [key ...]
    pub fn mget(&mut self, keys: &[&str]) {
        let mut args: Vec<&[u8]> = vec![b"MGET"];
        for k in keys {
            args.push(k.as_bytes());
        }
        self.command(&args);
    }

    /// EXISTS key [key ...]
    pub fn exists(&mut self, keys: &[&str]) {
        let mut args: Vec<&[u8]> = vec![b"EXISTS"];
        for k in keys {
            args.push(k.as_bytes());
        }
        self.command(&args);
    }

    /// DBSIZE
    pub fn dbsize(&mut self) {
        self.command(&[b"DBSIZE"]);
    }

    // ── TTL commands ────────────────────────────────────────────────────

    /// EXPIRE key seconds
    pub fn expire(&mut self, key: &str, seconds: i64) {
        self.command(&[b"EXPIRE", key.as_bytes(), seconds.to_string().as_bytes()]);
    }

    /// TTL key
    pub fn ttl(&mut self, key: &str) {
        self.command(&[b"TTL", key.as_bytes()]);
    }

    /// PERSIST key
    pub fn persist(&mut self, key: &str) {
        self.command(&[b"PERSIST", key.as_bytes()]);
    }

    // ── Hash commands ───────────────────────────────────────────────────

    /// HSET key field value
    pub fn hset(&mut self, key: &str, field: &str, value: &str) {
        self.command(&[b"HSET", key.as_bytes(), field.as_bytes(), value.as_bytes()]);
    }

    /// HGET key field
    pub fn hget(&mut self, key: &str, field: &str) {
        self.command(&[b"HGET", key.as_bytes(), field.as_bytes()]);
    }

    /// HDEL key field [field ...]
    pub fn hdel(&mut self, keys: &[&str]) {
        let mut args: Vec<&[u8]> = vec![b"HDEL"];
        for k in keys {
            args.push(k.as_bytes());
        }
        self.command(&args);
    }

    /// HGETALL key
    pub fn hgetall(&mut self, key: &str) {
        self.command(&[b"HGETALL", key.as_bytes()]);
    }

    /// HLEN key
    pub fn hlen(&mut self, key: &str) {
        self.command(&[b"HLEN", key.as_bytes()]);
    }

    // ── List commands ───────────────────────────────────────────────────

    /// RPUSH key elem [elem ...]
    pub fn rpush(&mut self, key: &str, elems: &[&str]) {
        let mut args: Vec<&[u8]> = vec![b"RPUSH", key.as_bytes()];
        for e in elems {
            args.push(e.as_bytes());
        }
        self.command(&args);
    }

    /// LPUSH key elem [elem ...]
    pub fn lpush(&mut self, key: &str, elems: &[&str]) {
        let mut args: Vec<&[u8]> = vec![b"LPUSH", key.as_bytes()];
        for e in elems {
            args.push(e.as_bytes());
        }
        self.command(&args);
    }

    /// LLEN key
    pub fn llen(&mut self, key: &str) {
        self.command(&[b"LLEN", key.as_bytes()]);
    }

    /// LRANGE key start stop
    pub fn lrange(&mut self, key: &str, start: i64, stop: i64) {
        self.command(&[b"LRANGE", key.as_bytes(), start.to_string().as_bytes(), stop.to_string().as_bytes()]);
    }

    /// LPOP key
    pub fn lpop(&mut self, key: &str) {
        self.command(&[b"LPOP", key.as_bytes()]);
    }

    /// RPOP key
    pub fn rpop(&mut self, key: &str) {
        self.command(&[b"RPOP", key.as_bytes()]);
    }

    /// LTRIM key start stop
    pub fn ltrim(&mut self, key: &str, start: i64, stop: i64) {
        self.command(&[b"LTRIM", key.as_bytes(), start.to_string().as_bytes(), stop.to_string().as_bytes()]);
    }

    // ── Raw command ─────────────────────────────────────────────────────

    /// Add a raw RESP command.
    pub fn command(&mut self, args: &[&[u8]]) {
        self.bufs.push(encode_command(args));
    }

    // ── Execute ─────────────────────────────────────────────────────────

    /// Send all buffered commands and read all replies.
    ///
    /// Returns a `PipelineResult` that can be indexed to extract
    /// individual replies.
    pub async fn execute(self) -> Result<PipelineResult, Error> {
        if self.bufs.is_empty() {
            return Err(Error::PipelineEmpty);
        }

        // Write all commands in one shot
        let total_len: usize = self.bufs.iter().map(|b| b.len()).sum();
        let mut flat = Vec::with_capacity(total_len);
        for buf in &self.bufs {
            flat.extend_from_slice(buf);
        }

        use tokio::io::AsyncWriteExt;
        self.stream.write_all(&flat).await?;
        self.stream.flush().await?;

        // Read all replies
        let mut reader = BufReader::new(&mut *self.stream);
        let mut replies = Vec::with_capacity(self.bufs.len());
        for _ in &self.bufs {
            replies.push(read_reply(&mut reader).await?);
        }

        Ok(PipelineResult(replies))
    }
}

/// Holds replies from a pipeline execution.
pub struct PipelineResult(Vec<Reply>);

impl PipelineResult {
    /// Number of replies.
    pub fn len(&self) -> usize {
        self.0.len()
    }

    /// True if no replies.
    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }

    /// Get raw reply at index.
    pub fn get(&self, index: usize) -> Option<&Reply> {
        self.0.get(index)
    }

    /// Interpret reply at `index` as a string (null → None).
    pub fn string(&self, index: usize) -> Result<Option<String>, Error> {
        self.0.get(index)
            .ok_or_else(|| Error::Unexpected(format!("pipeline index {index} out of range")))?
            .to_string_opt()
    }

    /// Interpret reply at `index` as i64.
    pub fn integer(&self, index: usize) -> Result<i64, Error> {
        self.0.get(index)
            .ok_or_else(|| Error::Unexpected(format!("pipeline index {index} out of range")))?
            .to_i64()
    }

    /// Interpret reply at `index` as OK status.
    pub fn ok(&self, index: usize) -> Result<(), Error> {
        self.0.get(index)
            .ok_or_else(|| Error::Unexpected(format!("pipeline index {index} out of range")))?
            .expect_ok()
    }

    /// Interpret reply at `index` as a string slice (for LRANGE, HKEYS, etc.)
    pub fn string_slice(&self, index: usize) -> Result<Vec<String>, Error> {
        let reply = self.0.get(index)
            .ok_or_else(|| Error::Unexpected(format!("pipeline index {index} out of range")))?;
        crate::resp::decode_string_slice(reply)
    }

    /// Interpret reply at `index` as a hash map (for HGETALL).
    pub fn string_map(&self, index: usize) -> Result<Vec<(String, String)>, Error> {
        let reply = self.0.get(index)
            .ok_or_else(|| Error::Unexpected(format!("pipeline index {index} out of range")))?;
        crate::resp::decode_string_map(reply)
    }
}
